flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobSubmitHandlerTest.java

```diff
@@ -71,20 +112,131 @@ public void testSerializationFailureHandling() throws Exception {
+			TestingUtils.defaultExecutor());
+
+		JobSubmitRequestBody request = new JobSubmitRequestBody(jobGraphFile.getFileName().toString(), Collections.emptyList(), Collections.emptyList());
+
+		try {
+			handler.handleRequest(new HandlerRequest<>(request, EmptyMessageParameters.getInstance(), Collections.emptyMap(), Collections.emptyMap(), Arrays.asList(jobGraphFile.toFile(), countExceedingFile.toFile())), mockGateway)
+				.get();
+		} catch (Exception e) {
+			ExceptionUtils.findThrowable(e, candidate -> candidate instanceof RestHandlerException && candidate.getMessage().contains("count"));
+		}
+	}
+
+	@Test
+	public void testFileHandling() throws Exception {
+		final String dcEntryName = "entry";
+
+		CompletableFuture<JobGraph> submittedJobGraphFuture = new CompletableFuture<>();
+		DispatcherGateway dispatcherGateway = new TestingDispatcherGateway.Builder()
+			.setBlobServerPort(blobServer.getPort())
+			.setSubmitFunction(submittedJobGraph -> {
+				submittedJobGraphFuture.complete(submittedJobGraph);
+				return CompletableFuture.completedFuture(Acknowledge.get());
+			})
+			.build();
+
+		GatewayRetriever<DispatcherGateway> gatewayRetriever = new TestGatewayRetriever(dispatcherGateway);
+
+		JobSubmitHandler handler = new JobSubmitHandler(
+			CompletableFuture.completedFuture("http://localhost:1234"),
+			gatewayRetriever,
+			RpcUtils.INF_TIMEOUT,
+			Collections.emptyMap(),
+			TestingUtils.defaultExecutor());
+
+		final Path jobGraphFile = TEMPORARY_FOLDER.newFile().toPath();
+		final Path jarFile = TEMPORARY_FOLDER.newFile().toPath();
+		final Path artifactFile = TEMPORARY_FOLDER.newFile().toPath();
+
+		final JobGraph jobGraph = new JobGraph();
+		// the entry that should be updated
+		jobGraph.addUserArtifact(dcEntryName, new DistributedCache.DistributedCacheEntry("random", false));
+		try (OutputStream fileOut = Files.newOutputStream(jobGraphFile)) {
+			try (ObjectOutputStream objectOut = new ObjectOutputStream(fileOut)) {
+				objectOut.writeObject(jobGraph);
+			}
+		}
+
+		JobSubmitRequestBody request = new JobSubmitRequestBody(
+			jobGraphFile.getFileName().toString(),
+			Collections.singletonList(jarFile.getFileName().toString()),
+			Collections.singleton(new JobSubmitRequestBody.DistributedCacheFile(dcEntryName, artifactFile.getFileName().toString())));
+
+		handler.handleRequest(new HandlerRequest<>(
+				request,
+				EmptyMessageParameters.getInstance(),
+				Collections.emptyMap(),
+				Collections.emptyMap(),
+				Arrays.asList(jobGraphFile.toFile(), jarFile.toFile(), artifactFile.toFile())), dispatcherGateway)
+			.get();
+
+		Assert.assertTrue("No JobGraph was submitted.", submittedJobGraphFuture.isDone());
+		final JobGraph submittedJobGraph = submittedJobGraphFuture.get();
+		Assert.assertEquals(1, submittedJobGraph.getUserJarBlobKeys().size());
+		Assert.assertEquals(1, submittedJobGraph.getUserArtifacts().size());
+		Assert.assertNotNull(submittedJobGraph.getUserArtifacts().get(dcEntryName).blobKey);
```

Comment:
Just a side note, hamcrest offers a bit more expressive assertions which generate in many cases better failure messages. E.g., `assertThat(submittedJobGraph.getUserArtifacts(), hasSize(1))`.